receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,539 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snmpreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/snmpreceiver"
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
+
+	"github.com/gosnmp/gosnmp"
+)
+
+// Custom errors
+var (
+	errNoGetOIDs        = errors.New(`all GET OIDs requests have failed`)
+	errNoProcessGetOIDs = errors.New(`all attempts to process GET OIDs have failed`)
+	errNoWalkOIDs       = errors.New(`all WALK OIDs requests have failed`)
+)
+
+// Trimmed down list of SNMP data types to make things a little more simple for the scraper
+type oidDataType byte
+
+const (
+	NotSupported oidDataType = 0x00
+	Integer      oidDataType = 0x01 // value will be int64
+	Float        oidDataType = 0x02 // value will be float64
+	String       oidDataType = 0x03 // value will be string
+)
+
+// snmpData used for processFunc and is a simpler version of gosnmp.SnmpPDU
+type snmpData struct {
+	parentOID string // optional
+	oid       string
+	value     interface{}
+	valueType oidDataType
+}
+
+// processFunc our own function type to better control what data can be passed in
+type processFunc func(data snmpData) error
+
+// client is used for retrieving data from a SNMP environment
+type client interface {
+	// GetScalarData retrieves SNMP scalar data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetScalarData(oids []string, processFn processFunc) error
+	// GetIndexedData retrieves SNMP indexed data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetIndexedData(oids []string, processFn processFunc) error
+	// Connect makes a connection to the SNMP host
+	Connect() error
+	// Close closes a connection to the SNMP host
+	Close() error
+}
+
+// snmpClient implements the client interface and retrieves data through SNMP
+type snmpClient struct {
+	client goSNMPWrapper
+	logger *zap.Logger
+}
+
+// Verify snmpClient implements client interface
+var _ client = (*snmpClient)(nil)
+
+// newClient creates an initialized client
+func newClient(cfg *Config, host component.Host, settings component.TelemetrySettings, logger *zap.Logger) (client, error) {
+	// Create goSNMP client
+	goSNMP := newGoSNMPWrapper()
+	goSNMP.SetTimeout(cfg.CollectionInterval)
+
+	// Set goSNMP version based on config
+	switch cfg.Version {
+	case "v3":
+		goSNMP.SetVersion(gosnmp.Version3)
+	case "v2c":
+		goSNMP.SetVersion(gosnmp.Version2c)
+	case "v1":
+		goSNMP.SetVersion(gosnmp.Version1)
+	default:
+		return nil, errors.New("failed to create goSNMP client: invalid version")
```

Comment:
Should we ever hit this error? Assuming this would be vetted at the config level?

Reply:
@cpheps nope. like i mentioned elsewhere i think a fair amount of this logic is doing double checks and ignores the fact that the config is validating up front. i was a little torn on whether i should have that frame of mind or not, and in the end i only did a half baked validation here in the client which is probably the worst of both worlds. i have no super strong opinions here and can remove any duplicate validation if you think it's warranted

Reply:
@cpheps i went ahead and removed any checks in the client (and client tests) that would already be covered in config validation